### ISSUE TITLE
Cannot choose custom allocator

### ISSUE THREAD

**Issue by shahn** [rust-fuzz/cargo-fuzz#104]
Anything using a custom allocator cannot be fuzzed, even if the same allocator is chosen in the fuzzer script.

**Comment by Manishearth** [comment]
Could you provide more information about the errors you see?

**Comment by shahn** [comment]
Of course, here we go for the various combinations:

rlib + bin, bin + rlib, staticlib + bin, bin + staticlib:

  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/fuzz/target/x86_64-unknown-linux-gnu/debug/deps/consdiff-876121c7e733b4a7.0.o" "-o" "/fuzz/target/x86_64-unknown-linux-gnu/debug/deps/consdiff-876121c7e733b4a7" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/fuzz/target/x86_64-unknown-linux-gnu/debug/deps" "-L" "/fuzz/target/debug/deps" "-L" "../../common" "-L" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "or-ctime" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-l" "or-testing" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "/fuzz/target/x86_64-unknown-linux-gnu/debug/deps/liblibc-5dc7b85e748840b4.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-f4594d3e53dcb114.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-1efbcfd8938372b6.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-532a3dbf317eff86.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_unicode-cfbd6648f7db2ee5.rlib" "/fuzz/target/x86_64-unknown-linux-gnu/debug/deps/libtiny_keccak-fbf86ba37ae3101c.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-488b4ab4bd53a138.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-ca07b617414dd0fa.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-492d8ea7fa3384ff.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_abort-6800222f9a60165a.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-88c194c15fdb6521.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-687e6a964d22cbb4.rlib" "/home/shahn/.rustup/toolchains/nightly-2017-03-28-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-987729be881d4d32.rlib" "-l" "util" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "util"

  = note: /fuzz/target/x86_64-unknown-linux-gnu/debug/deps/consdiff-876121c7e733b4a7.0.o: In function `core::slice::{{impl}}::next_back<usize>':
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__sanitizer_cov'
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__asan_option_detect_stack_use_after_return'
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__sanitizer_cov'
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__sanitizer_cov'
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__asan_stack_malloc_0'
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__sanitizer_cov'
          /checkout/src/libcore/slice/mod.rs:1165: undefined reference to `__sanitizer_cov'
          /checkout/src/libcore/slice/mod.rs:1169: undefined reference to `__asan_report_load8'
          /checkout/src/libcore/slice/mod.rs:1168: undefined reference to `__sanitizer_cov'
          /checkout/src/libcore/slice/mod.rs:1170: undefined reference to `__asan_report_load8'
          /checkout/src/libcore/slice/mod.rs:1172: undefined reference to `__asan_report_load8'
          /checkout/src/libcore/slice/mod.rs:1172: undefined reference to `__asan_report_load8'
          ... another 1.3MB of errors

staticlib + rlib:
     Running `rustc --crate-name consdiff lib.rs --crate-type staticlib --crate-type rlib --e
	 error: Only executables and rlibs can be compiled with `-Z sanitizer`

rlib, rlib + staticlib: works

**Comment by shahn** [comment]
Errors I get are these:

"cannot link together two allocators: custom_alloc and alloc_system". This happens even when I say "extern crate custom_alloc;" in the fuzzer_script. Somehow it always wants to pull in alloc_system

**Comment by dwrensha** [comment]
Yeah, I see this error when I try to fuzz [miri](https://github.com/solson/miri). I assume it has something to do with `#![feature(rustc_private)]`.

**Comment by whitequark** [comment]
I think the new global_allocator attribute (RFC 1974) took care of it. Can you recheck?

**Comment by oech3** [comment]
I see many SEGV at uutils/coreutils with mimalloc e.g. https://github.com/uutils/coreutils/actions/runs/22096265250/job/63854290831?pr=10868 .
Using `mimalloc = { version = "0.1.48", default-features = false, features = [
  "override",
] }`. I cannot remove `override` for performance.
